### PR TITLE
Feature/l6 c add contacts form

### PR DESCRIPTION
**Pull request by Khikar-1997**

### CODE REVIEW

File: src/Actions/contacts.js [hunk 1]
```diff
@@ -0,0 +1,3 @@
+import { createAction } from '@reduxjs/toolkit/src/createAction';
```
Comment (AnushavanGhulyan) [inline src/Actions/contacts.js]: please fix import path

File: src/Actions/contacts.js [hunk 2]
```diff
@@ -0,0 +1,3 @@
...
+
+export const ADD_CONTACT = createAction('CONTACT/ADD_CONTACT');
```
Comment (AnushavanGhulyan) [inline src/Actions/contacts.js]: make actions lowerCase

File: src/Components/AddContacts/AddContact.js [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+    if (name && phoneNumber) {
+      dispatch(
+        ADD_CONTACT({
+          id,
+          name,
+          surname,
+          avatar_image: '',
+          phone_number: phoneNumber,
+          favorites: false,
+        }),
```
Comment (AnushavanGhulyan) [inline src/Components/AddContacts/AddContact.js]: remove Id generation from here.
do it inside reducer.

see example [https://redux-toolkit.js.org/api/createAction#using-prepare-callbacks-to-customize-action-contents](https://redux-toolkit.js.org/api/createAction#using-prepare-callbacks-to-customize-action-contents),

use nanoId or your integrated uuid there

File: src/Components/ContactList/FavoritesList.js [hunk 4]
```diff
@@ -1,12 +1,15 @@
...
 import Contact from './Contact';
 
 import filterFavorites from '../../Utils/filterFavorites';
+import { getContacts } from '../../Selectors/contacts';
 
-const FavoritesList = ({ data }) => {
-  const favoritesList = useMemo(() => filterFavorites(data), [data]);
+const FavoritesList = () => {
+  const contacts = useSelector(getContacts);
+  const favoritesList = useMemo(() => filterFavorites(contacts), [contacts]);
```
Comment (AnushavanGhulyan) [inline src/Components/ContactList/FavoritesList.js]: create selector for filterFavorites based on getContacts. 
see example [https://redux-toolkit.js.org/api/createSelector](https://redux-toolkit.js.org/api/createSelector)

File: src/Components/ContactList/List.js [hunk 5]
```diff
@@ -1,14 +1,17 @@
...
 import contactListWithTitle from '../../Utils/contactListWithTitle';
+import { getContacts } from '../../Selectors/contacts';
 
 import colors from '../../Constants/colors';
 
-const List = ({ data }) => {
-  const contactList = useMemo(() => contactListWithTitle(data), [data]);
+const List = () => {
+  const contacts = useSelector(getContacts);
+  const contactList = useMemo(() => contactListWithTitle(contacts), [contacts]);
```
Comment (AnushavanGhulyan) [inline src/Components/ContactList/List.js]: same here

File: src/Components/ContactSearch/SearchedContacts.js [hunk 6]
```diff
@@ -1,21 +1,24 @@
...
 
 const SearchedContacts = ({ contacts }) => (
-  <FlatList
-    data={contacts}
-    renderItem={({ item }) => <Contact contact={item} />}
-    keyExtractor={item => item.id}
-    contentInsetAdjustmentBehavior="automatic"
-    style={styles.searchedContactsList}
-  />
+  <View style={{ flex: 1 }}>
```
Comment (AnushavanGhulyan) [inline src/Components/ContactSearch/SearchedContacts.js]: remove inline styles

File: src/Config/store.js [hunk 7]
```diff
@@ -0,0 +1,4 @@
+import { createStore } from '@reduxjs/toolkit';
+import contactsReducer from '../Reducers/contacts';
+
+export default createStore(contactsReducer);
```
Comment (AnushavanGhulyan) [inline src/Config/store.js]: change to configure store.

[https://redux-toolkit.js.org/tutorials/quick-start#create-a-redux-store](https://redux-toolkit.js.org/tutorials/quick-start#create-a-redux-store)

File: src/Reducers/contacts.js [hunk 8]
```diff
@@ -0,0 +1,13 @@
+import contacts from '../../data/MOCK_DATA.json';
+import { ADD_CONTACT } from '../Actions/contacts';
+
+const initialState = {
+  value: contacts,
+};
```
Comment (AnushavanGhulyan) [inline src/Reducers/contacts.js]:
```suggestion
const initialState = {
 contacts: MOCK_DATA,
};
```

File: src/Screens/AddContact.screen.js [hunk 9]
```diff
@@ -1,11 +1,23 @@
...
 
 import * as Routes from '../Constants/navigationRouts';
+import colors from '../Constants/colors';
 
-const AddContactScreen = () => <View />;
+const AddContactScreen = () => (
+  <View style={styles.container}>
+    <AddContact />
+  </View>
+);
```
Comment (AnushavanGhulyan) [inline src/Screens/AddContact.screen.js]: is this View is necessary?

File: src/Components/AddContacts/AddContact.js [hunk 10]
```diff
@@ -0,0 +1,77 @@
+import React, { useState } from 'react';
```
Comment (AnushavanGhulyan) [inline src/Components/AddContacts/AddContact.js]: make attention to file names.

_src/Components/AddContacts/index.js_ is better name

**Review by AnushavanGhulyan — CHANGES_REQUESTED** [review]

File: src/Components/AddContacts/index.js [hunk 11]
```diff
@@ -0,0 +1,73 @@
...
+    }
+  };
+
+  return (
+    <View style={styles.container}>
+      <FiledForm
+        value={name}
+        placeholder="name"
+        autoCapitalize="sentences"
+        placeholderTextColor={colors.logan}
```
Comment (griarman) [inline src/Components/AddContacts/index.js]: please move this style into FiledForm

File: src/Components/AddContacts/index.js [hunk 12]
```diff
@@ -0,0 +1,73 @@
...
+        icon="profile"
+        onChangeText={setName}
+      />
+      <FiledForm
+        value={surname}
+        placeholder="surname"
+        autoCapitalize="sentences"
+        placeholderTextColor={colors.logan}
```
Comment (griarman) [inline src/Components/AddContacts/index.js]: same here

File: src/Components/AddContacts/index.js [hunk 13]
```diff
@@ -0,0 +1,73 @@
...
+        icon="profile"
+        onChangeText={setSurname}
+      />
+      <FiledForm
+        icon="call"
+        value={phoneNumber}
+        placeholder="phone"
+        keyboardType="phone-pad"
+        placeholderTextColor={colors.logan}
```
Comment (griarman) [inline src/Components/AddContacts/index.js]: same here

File: src/Reducers/contacts.js [hunk 14]
```diff
@@ -0,0 +1,13 @@
+import contacts from '../../data/MOCK_DATA.json';
+import { addContact } from '../Actions/contacts';
+
+const initialState = {
+  contacts: contacts,
```
Comment (griarman) [inline src/Reducers/contacts.js]: just write contacts

File: src/Screens/AddContact.screen.js [hunk 15]
```diff
@@ -1,11 +1,11 @@
 import React from 'react';
-import { View } from 'react-native';
 
 import Button from '../UIKit/Button';
+import AddContact from '../Components/AddContacts';
 
 import * as Routes from '../Constants/navigationRouts';
 
-const AddContactScreen = () => <View />;
+const AddContactScreen = () => <AddContact />;
```
Comment (griarman) [inline src/Screens/AddContact.screen.js]: let instead of returning AddContact component move AddContact data into this screen

**Review by griarman — CHANGES_REQUESTED** [review]

File: src/Actions/contacts.js [hunk 16]
```diff
@@ -0,0 +1,15 @@
+import { createAction } from '@reduxjs/toolkit';
+
+import 'react-native-get-random-values';
+import { v4 as uuidv4 } from 'uuid';
+const id = uuidv4();
```
Comment (AnushavanGhulyan) [inline src/Actions/contacts.js]: this will work only one time. so all new IDs will the same.
move Id creation inside action

**Review by AnushavanGhulyan — CHANGES_REQUESTED** [review]

**Review by AnushavanGhulyan — APPROVED** [review]

File: src/Reducers/contacts.js [hunk 17]
```diff
@@ -0,0 +1,35 @@
...
+import 'react-native-get-random-values';
+import { v4 as uuidv4 } from 'uuid';
+
+import contacts from '../../data/MOCK_DATA.json';
+
+const initialState = {
+  contacts,
+};
+
+export const counterSlice = createSlice({
```
Comment (AnushavanGhulyan) [inline src/Reducers/contacts.js]: Please rename

**Review by AnushavanGhulyan — APPROVED** [review]

File: src/Screens/SearchContact.screen.tsx [hunk 18]
```diff
@@ -42,7 +52,22 @@ const SearchContact = () => {
...
+  );
+};
+
+type FunctionWithOptions = {
+  options: (props: {
+    route: any;
+    navigation: any;
+  }) => NativeStackNavigationOptions;
+  (): ReactElement;
 };
```
Comment (AnushavanGhulyan) [inline src/Screens/SearchContact.screen.tsx]: Let's declare this type global and use it for other screens also

**Review by AnushavanGhulyan — CHANGES_REQUESTED** [review]

File: src/Components/AddContacts/ImageUploadForm.tsx [hunk 19]
```diff
@@ -0,0 +1,32 @@
...
+  ImageURISource,
+  GestureResponderEvent,
+} from 'react-native';
+
+export const ImageUploadForm = ({
+  pickPicture,
+  uri,
+  source,
+}: {
+  pickPicture: (event: GestureResponderEvent) => void;
```
Comment (AnushavanGhulyan) [inline src/Components/AddContacts/ImageUploadForm.tsx]: wrong event type

File: src/Screens/AddContact.screen.tsx [hunk 20]
```diff
@@ -0,0 +1,118 @@
...
+import { ImageUploadForm } from '../Components/AddContacts/ImageUploadForm';
+
+import colors from '../Constants/colors';
+import { CONTACTS } from '../Constants/navigationRouts';
+import { addContact } from '../Reducers/contacts';
+import imagePaths from '../Constants/imagePaths';
+
+interface AvatarProps {
+  onChange?: (image: ImageOrVideo) => void;
+  source: ImageURISource;
```
Comment (AnushavanGhulyan) [inline src/Screens/AddContact.screen.tsx]: is this the right interface for this screen?

**Review by AnushavanGhulyan — CHANGES_REQUESTED** [review]

File: src/Screens/AddContact.screen.tsx [hunk 21]
```diff
@@ -0,0 +1,118 @@
...
+    }
+  };
+
+  return (
+    <View style={styles.container}>
+      <View style={styles.uploadImageContainer}>
+        <ImageUploadForm
+          source={imagePaths.placeholderImage}
+          uri={uri}
+          props={props}
```
Comment (AnushavanGhulyan) [inline src/Screens/AddContact.screen.tsx]: remove props

File: src/Screens/AddContact.screen.tsx [hunk 22]
```diff
@@ -0,0 +1,118 @@
...
+      );
+      return navigation.navigate(CONTACTS);
+    }
+  };
+
+  return (
+    <View style={styles.container}>
+      <View style={styles.uploadImageContainer}>
+        <ImageUploadForm
+          source={imagePaths.placeholderImage}
```
Comment (AnushavanGhulyan) [inline src/Screens/AddContact.screen.tsx]: remove source

File: src/Components/AddContacts/ImageUploadForm.tsx [hunk 23]
```diff
@@ -0,0 +1,32 @@
...
+  uri: string;
```
Comment (AnushavanGhulyan) [inline src/Components/AddContacts/ImageUploadForm.tsx:11]: remove uri props.
use source instead

File: src/Screens/AddContact.screen.tsx (hunk 22 above)
Comment (AnushavanGhulyan) [inline src/Screens/AddContact.screen.tsx]:
```suggestion
          source={uri ? { uri } :  imagePaths.placeholderImage}
```

File: src/Screens/AddContact.screen.tsx [hunk 24]
```diff
@@ -0,0 +1,118 @@
...
+import { CONTACTS } from '../Constants/navigationRouts';
+import { addContact } from '../Reducers/contacts';
+import imagePaths from '../Constants/imagePaths';
+
+interface AvatarProps {
+  onChange?: (image: ImageOrVideo) => void;
+  source: ImageURISource;
+}
+
+const AddContactScreen = (props: AvatarProps) => {
```
Comment (AnushavanGhulyan) [inline src/Screens/AddContact.screen.tsx]: remove "props"

**Review by AnushavanGhulyan — APPROVED** [review]

**Pull request merged**